retrievalmarket/impl/providerstates/provider_states.go

```diff
@@ -30,7 +30,23 @@ type ProviderDealEnvironment interface {
 }
 
 func firstSuccessfulUnseal(ctx context.Context, node rm.RetrievalProviderNode, pieceInfo piecestore.PieceInfo) (io.ReadCloser, error) {
+
+	// prefer an unsealed sector containing the piece if one exists
+	for _, deal := range pieceInfo.Deals {
+		isUnsealed, err := node.IsUnsealed(ctx, deal.SectorID, deal.Offset.Unpadded(), deal.Length.Unpadded())
+		if err != nil {
+			continue
+		}
+		if isUnsealed {
+			reader, err := node.UnsealSector(ctx, deal.SectorID, deal.Offset.Unpadded(), deal.Length.Unpadded())
```

Comment:
I didn't get this the first time I looked at it - let's add a comment explaining that UnsealSector will get the data from the cache if it's already unsealed